// @qio/core/lib/runtimes/TestRuntime.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.testRuntime = exports.TestRuntime = void 0;
const prelude_1 = require("@qio/prelude");
const ts_scheduler_1 = require("ts-scheduler");
const FiberConfig_1 = require("../internals/FiberConfig");
const Exit_1 = require("../main/Exit");
const FiberRuntime_1 = require("./FiberRuntime");
class TestRuntime extends FiberRuntime_1.FiberRuntime {
    constructor(scheduler, config) {
        super();
        this.scheduler = scheduler;
        this.config = config;
    }
    configure(config) {
        return new TestRuntime(this.scheduler, config);
    }
    unsafeExecuteSync(io) {
        const exit = this.unsafeExecuteSync0(io);
        return exit !== undefined
            ? Exit_1.Exit.fold(exit)(undefined, prelude_1.Id, prelude_1.Id)
            : undefined;
    }
    unsafeExecuteSync0(io) {
        let result;
        this.unsafeExecute(io, (_) => (result = _));
        this.scheduler.run();
        return result;
    }
}
exports.TestRuntime = TestRuntime;
exports.testRuntime = (scheduler = ts_scheduler_1.testScheduler()) => new TestRuntime(scheduler, FiberConfig_1.FiberConfig.DEFAULT);
