// @qio/core/lib/internals/PureMutableList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PureMutableList = void 0;
const standard_data_structures_1 = require("standard-data-structures");
const QIO_1 = require("../main/QIO");
class PureMutableList {
    constructor() {
        this.list = standard_data_structures_1.DoublyLinkedList.of();
    }
    static of() {
        return QIO_1.QIO.lift(() => new PureMutableList());
    }
    get asArray() {
        return QIO_1.QIO.lift(() => this.list.asArray);
    }
    get isEmpty() {
        return QIO_1.QIO.lift(() => this.list.isEmpty);
    }
    get length() {
        return QIO_1.QIO.lift(() => this.list.length);
    }
    get shift() {
        return QIO_1.QIO.lift(() => this.list.shift());
    }
    add(element) {
        return QIO_1.QIO.lift(() => this.list.add(element));
    }
    forEach(f) {
        const itar = () => this.shift.chain((_) => _.map(f).getOrElse(QIO_1.QIO.void()).chain(itar));
        return itar();
    }
    remove(node) {
        return QIO_1.QIO.lift(() => this.list.remove(node));
    }
}
exports.PureMutableList = PureMutableList;
