// @qio/core/lib/main/Exit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Exit = exports.FiberExitTag = void 0;
var FiberExitTag;
(function (FiberExitTag) {
    FiberExitTag[FiberExitTag["SUCCESS"] = 0] = "SUCCESS";
    FiberExitTag[FiberExitTag["FAILURE"] = 1] = "FAILURE";
    FiberExitTag[FiberExitTag["CANCELLED"] = 2] = "CANCELLED";
})(FiberExitTag = exports.FiberExitTag || (exports.FiberExitTag = {}));
class FiberCancelled {
    constructor() {
        this.tag = FiberExitTag.CANCELLED;
    }
}
class FiberFailure {
    constructor(cause) {
        this.cause = cause;
        this.tag = FiberExitTag.FAILURE;
    }
}
class FiberSuccess {
    constructor(value) {
        this.value = value;
        this.tag = FiberExitTag.SUCCESS;
    }
}
exports.Exit = {
    ...FiberExitTag,
    cancel: () => new FiberCancelled(),
    fail: (cause) => new FiberFailure(cause),
    fold: (exit) => (S, AA, EE) => {
        switch (exit.tag) {
            case FiberExitTag.CANCELLED:
                return S;
            case FiberExitTag.FAILURE:
                return EE(exit.cause);
            case FiberExitTag.SUCCESS:
                return AA(exit.value);
        }
    },
    succeed: (value) => new FiberSuccess(value),
};
